c/datatable.h

```diff
@@ -183,4 +209,30 @@ RowIndex natural_join(const DataTable* xdt, const DataTable* jdt);
 
 //==============================================================================
 
+template <void (*F1)(size_t i, size_t j),
+          void (*FN)(size_t i, const intvec& js)>
+void DataTable::iterate_rows(size_t row0, size_t row1) {
+  if (row1 == size_t(-1)) {
+    row1 = nrows;
+  }
+  std::vector<RowColIndex> rcs = split_columns_by_rowindices();
```

Comment:
Btw, what about storing this `rcs` as a property in `DataTable`, and invalidate it when there are some relevant changes to the columns? Then, if several passes over the data is needed (for example when we do `nepochs > 1` in FTRL), `rcs` won't have to be calculated multiple times.

Reply:
Yes, I can imagine there are many places in the code where RowIndex could be affected. By the way, when we had `RowIndex` as a `DataTable` property, how this one was changed when one did `column->reify()`? 

I guess, `rcs` could also be an optional parameter to `iterate_rows`, if provided it should be used as is and not re-calculated. Actually, for FTRL I store `RowIndex` for each column within a corresponding hasher object, and then apply it while hashing. Do you think this is better/worse in terms of performance comparing this this new iterator?

Reply:
With your approach there is a cost of one virtual function call per column. With my approach there is a cost of copying the value `j` into once per column. Alternatively this could be implemented as double lookup.